[2단계 - 로또 리팩토링] 강산(조재중) 미션 제출합니다.

m-a-king opened the pull request:
> 안녕하세요, 파랑~ 😊
> 
> 좋은 리뷰 남겨주셔서 정말 감사합니다!
> 이번 PR에서는 다음과 같은 개선을 진행했습니다.
> 
> 1. 로또 번호를 일급 객체로 래핑하여 관리 
> 2. 로또 번호 생성 로직의 관점을 변경하여 비즈니스 로직 내에서 처리  
> 3. 로또 규칙이 변경되더라도 유연하게 적용될 수 있도록 설계  
> 
> 리뷰 부탁드립니다! 감사합니다.

___

안녕하세요! 파랑, 리뷰에서 요청해 주신 사항을 반영했습니다.
그 외에도 Money 클래스에서 로또와 관련된 책임이 있는 것을 분리하고, 테스트 코드를 조금 더 유연하게 작동할 수 있도록 리팩토링했습니다.

감사합니다~

m-a-king committed:
feat(LottoNumber): 로또 번호를 다루는 클래스 구현

- 로또 번호 객체 캐싱(동일한 값의 로또 번호는 동일한 객체를 참조하도록 변경).
- 로또 번호 범위 예외 처리를 LottoNumber 내부로 이동.

m-a-king committed:
refactor(LottoNumber): 원시값(int) 대신 LottoNumber 객체로 감싸도록 변경

- 로또 번호를 원시값(int)으로 관리하던 방식을 객체로 래핑
- 일관된 데이터 검증 및 불변성을 유지하도록 개선

m-a-king committed:
refactor: 가독성 개선, 명명 수정

m-a-king committed:
feat(Lotto, Lottos): 불변성 보장

- Lotto:
  - 생성 시 입력된 리스트를 `.stream().sorted().toList()`로 복사하여 정렬된 불변 리스트로 저장

- Lottos:
  - 생성자에서 `List.copyOf(lottos)` 사용하여 리스트 복사 및 불변 리스트로 저장

- LottoNumber:
  - 미리 생성된 객체 캐싱으로 동일한 숫자의 LottoNumber는 같은 객체를 참조하도록 변경

m-a-king committed:
refactor(Money): 정수 인자를 받아 Money 객체 생성하도록 변경

m-a-king committed:
feat(LottoRule): 로또 규칙 검증 로직 추가

- LottoRule 값들이 유효한 범위인지 정적 메서드로 검증
- 최소/최대 로또 번호, 선택 개수, 가격이 올바르게 설정되었는지 확인

m-a-king committed:
refactor: 메서드 명명 수정, 깃허브 마지막 줄 개행 문제 해결

summerlunaa commented on src/main/java/domain/Lotto.java:
```diff
@@ -2,61 +2,54 @@
 
-        numbers.sort(Comparator.naturalOrder());
-        this.numbers = numbers;
+    public static Lotto from(List<LottoNumber> numbers) {
+        return new Lotto(numbers);
     }
```
기존의 생성자와의 차이점은 무엇인가요?

m-a-king replied:
굳이 정적 팩토리 메서드를 활용할 필요가 없지만, 일관성있는 객체 생성 형태를 위해서 Lotto 객체에서도 이렇게 진행했습니다.

summerlunaa commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,57 @@
+                            LottoRule.MIN_LOTTO_NUMBER.getValue(),
+                            LottoRule.MAX_LOTTO_NUMBER.getValue())
+                    .boxed()
+                    .collect(Collectors.toMap(
+                            key -> key,
+                            LottoNumber::new));
```
```suggestion
                    .collect(Collectors.toMap(key -> key, LottoNumber::new));
```

어느 정도 취향의 영역이긴 하지만 과한 줄바꿈도 가독성을 해칠 수 있습니다. 전체적으로 점검해주세요.

m-a-king replied:
감사합니다. 고쳐주신 코드가 더 가독성이 좋은 것 같아요~

summerlunaa commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,57 @@
+
+public class LottoNumber {
+
+    private static final Map<Integer, LottoNumber> CACHE =
```
캐싱까지 하셨네요 👍  어떤 장점이 있나요?

m-a-king replied:
로또를 아무리 많이 생성해도, 로또 번호 객체는 설정된 최솟값부터 최댓값까지만 생성되기 때문에, 불필요한 객체 생성을 방지하고 메모리 사용량을 줄일 수 있습니다. 또한, 동일한 객체를 재사용하기 때문에 GC 부담이 줄어들어 성능적인 이점이 있을 것으로 예상됩니다!

summerlunaa commented on src/main/java/domain/LottoRanking.java:
```diff
@@ -22,7 +22,7 @@ public enum LottoRanking {
     }
 
-    public static LottoRanking from(int correctCount, boolean isCorrectBonusNumber) {
+    public static LottoRanking of(int correctCount, boolean isCorrectBonusNumber) {
```
리뷰 안 드렸는데도 잘 변경하셨군요 👍

m-a-king replied:
네! 인지했을 때, from으로 두고 correctNumber와 isCorrectBonusNumber를 래핑하려고 했는데 of로 변경하는 것이 더 직관적인 것 같아서 이렇게 진행했습니다!

summerlunaa commented on src/main/java/domain/LottoRule.java:
```diff
@@ -16,7 +16,25 @@ public int getValue() {
 
-    public static boolean isLottoRange(int number) {
-        return number >= MIN_LOTTO_NUMBER.value && number <= MAX_LOTTO_NUMBER.value;
+    static {
+        validateRange();
+        validateCount();
+    }
```
이 검증들은 왜 필요한가요?

m-a-king replied:
개발자의 실수로 로또 규칙 자체를 잘못 설정하면, 프로그램 실행 시점에 즉시 확인할 수 있도록 했습니다~

summerlunaa commented on src/main/java/domain/Lottos.java:
```diff
@@ -1,25 +1,23 @@
 
-    public void addLotto(Lotto lotto) {
-        lottos.add(lotto);
+    public int getLottoCount() {
+        return lottos.size();
     }
```
setter를 제거하고 불변객체로 만드셨네요 👍

summerlunaa commented on src/main/java/domain/Money.java:
```diff
@@ -32,12 +34,4 @@ private static void validateZeroAmount(int amount) {
-            throw new IllegalArgumentException("구입 금액은 숫자여야 합니다.");
-        }
-    }
-}
```
<img width="111" alt="image" src="https://github.com/user-attachments/assets/c6993f23-a2be-40d8-8123-686084e5b263" />

어떤 걸 의미하는 표시일까요?

m-a-king replied:
POSIX 표준으로 알고 있습니다. 당장 문제가 없어도 잠재적인 문제를 방지하기 위해서 개행 추가하겠습니다!

summerlunaa commented on src/main/java/service/LottoMachine.java:
```diff
@@ -1,29 +1,20 @@
 package service;
```
LottoMachine, Generator, Checker는 왜 service 패키지에 두셨나요? domain과 무엇이 다른가요?

m-a-king replied:
이전에 상태를 갖지 않고 도메인을 호출하는 클래스들을 서비스라고 칭했는데, 현 상황에서는 모두 도메인으로 옮기는 방법이 더 적절할 것 같아요. 맞을까요??

summerlunaa replied:
넵 그런 이유라면 도메인이 적절해보입니다.

summerlunaa commented on src/main/java/service/LottoMachine.java:
```diff
@@ -1,29 +1,20 @@
-            lottos.addLotto(lotto);
-        }
+    public Lottos buyLottos(Money money) {
+        int ticket = money.getAmount() / LottoRule.LOTTO_PRICE.getValue();
```
변수명이 조금 더 명확하면 좋을 것 같아요. 구입한 로또의 개수라는 걸 더 잘 나타낼 수 있게요!

m-a-king replied:
티켓이라는 변수명이 좋다고 생각했는데, 누군가에게는 명확해 보이지 않을 수 있기에 직관적으로 명명하는 게 더 좋다는 것을 알았습니다..

summerlunaa commented on src/main/java/service/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private void validateRepeat(int repeatCount) {
+        int minRepeatCount = 1;
```
상수로 정의할까요?

m-a-king replied:
상수로 정의해도 좋을 것 같습니다! 그렇지만 하나의 메서드에서만 사용되고 그 메서드가 검증을 다룬다면 메서드 내의 변수로 관리하는 정도로 괜찮지 않은지 여쭤보고 싶습니다. 매직 넘버로 1이라고 적었다면 분명히 수정해야겠지만 이런 디테일이 궁금해서 파랑에게 여쭤보고 싶어요!

m-a-king replied:
라고 생각했는데, 테스트에서 위 변수를 사용할 수 없을 것 같아서 상수로 빼는 것이 더 좋을 것 같습니다! 감사합니다.

summerlunaa replied:
좋습니다. 값이 해당 메서드 내에서만 사용된다면 변수로 정의해두어도 괜찮아보여요 :) 다른 곳에서도 사용된다면 상수로 빼는 것을 추천드립니다.

summerlunaa commented on src/test/java/service/LottoNumberGeneratorTest.java:
```diff
@@ -0,0 +1,92 @@
+
+    @Test
+    @DisplayName("요청된 개수만큼 로또를 생성해야 한다")
+    void 로또_개수만큼_생성되어야_한다() {
```
숫자가 랜덤이라 로직이 잘못되었더라도 어떨 땐 성공하고 어떨 땐 실패할 것 같은데 LottoNumberGeneratorTest에 대한 강산의 생각이 궁금합니다 ㅎㅎ

m-a-king replied:
숫자가 랜덤이라는 것은 repeatCount가 5가 아니라 21억이나 -1이라면 문제가 발생하니 테스트 명이 조금 더 명확해져야 한다는 걸까요?

summerlunaa replied:
아뇨 Generator에서 로또 숫자를 랜덤으로 생성하기 때문에 1~45 사이의 숫자를 잘 생성하는지를 확인하거나 중복을 확인할 때 테스트의 결과를 신뢰할 수 있냐는 의미였어요.

m-a-king replied:
이해했습니다!

랜덤 값 생성 자체를 검증하려면 사실 가짜 데이터를 만들어야 하는데, 일반적으로 모킹을 하거나, 전략 패턴을 이용해서 고정된 값을 반환하는 테스트 전용 클래스를 사용하게 됩니다. 하지만 이렇게 하면 결국 실제 랜덤성을 검증하는 것이 아니라, 정해진 값에 대한 테스트가 될 가능성이 높아 테스트의 충실도가 낮아질 수도 있다고 생각했습니다.

그래서 저는 관점을 조금 다르게 가져가 봤습니다.
랜덤 값이 올바르게 생성되는지를 직접 테스트하기보다는, 생성된 값이 비즈니스 로직을 망가뜨리지 않는지를 검증하는 것이 더 의미 있는 방향이 아닐까 싶었습니다.
즉, LottoNumber 클래스에서 1~45 범위를 벗어나는 값을 검증하는 방식으로 책임을 구분하면, 별도로 랜덤 값 생성을 테스트할 필요 없이 비즈니스 로직을 안정적으로 보호할 수 있을 것 같습니다.

특히, 로또에서는 랜덤 생성 자체도 비즈니스 로직의 일부이기 때문에, 이를 단순히 분리할 수도 없고, 무조건 테스트 대상에서 배제할 수도 없다고 생각합니다.
그래서 랜덤 생성과 로또 규칙 사이에 적절한 격벽을 두면서도, 로또의 핵심 개념을 유지할 수 있도록 설계했습니다.

이런 방향으로 접근하면 랜덤성 자체를 테스트하려는 어려움을 피하면서도, 비즈니스 로직의 신뢰성을 확보하는 데 집중할 수 있을 것 같았어요.

혹시 다른 의견 있으시면 알려주시면 감사하겠습니다!

summerlunaa commented on src/test/java/domain/MoneyTest.java:
```diff
@@ -0,0 +1,21 @@
+
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+class MoneyTest {
```
Money에 있는 검증에 대한 테스트들이 누락된 것 같아요.

m-a-king replied:
감사합니다. 채우겠습니다!

summerlunaa commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,33 @@
+        // given
+        // when
+        // then
+        assertThrows(IllegalArgumentException.class, () -> LottoNumber.from(LottoRule.MIN_LOTTO_NUMBER.getValue() - 1));
+        assertThrows(IllegalArgumentException.class, () -> LottoNumber.from(LottoRule.MAX_LOTTO_NUMBER.getValue() + 1));
```
```suggestion
        assertAll(
                () -> assertThrows(IllegalArgumentException.class, () -> LottoNumber.from(LottoRule.MIN_LOTTO_NUMBER.getValue() - 1)),
                () -> assertThrows(IllegalArgumentException.class, () -> LottoNumber.from(LottoRule.MAX_LOTTO_NUMBER.getValue() + 1))
        );
```

현재 코드와 assertAll을 사용했을 때의 코드의 차이점은 무엇일까요?

m-a-king replied:
확인해 보니 저번 리뷰에서 남겨주신 것과 비슷한 문제가 있는 것 같습니다. 변경 전은 첫 테스트 실패 시 이후의 테스트를 무시하고, assertAll로 감싸니 두 테스트 모두 실행 후 `Multiple Failures (2 failures)`이라는 키워드로 모든 테스트가 실패했음을 알려주는 것 같습니다!

감사합니다. 많이 배웠습니다!

summerlunaa reviewed (changes requested):
강산 안녕하세요~ 파랑입니다 🐳
2단계도 MVC 패턴 기반으로 잘 구성해주셨어요 👍 테스트도 꼼꼼히 작성해주셨구요. 리뷰 확인해보시고 궁금한 점은 코멘트나 DM으로 편하게 말씀해주세요!
마지막까지 파이팅입니다 💪

m-a-king committed:
refactor: 리뷰 반영

- 서비스 패키지 내 항목을 도메인 패키지로 이동
- 메서드 명, 변수 명, 코드 가독성 개선
- 최소 반복 횟수를 상수로 추출
- 다중 검증 시 assertAll 사용

m-a-king committed:
feat(Money): Money의 로또 관련 책임을 LottoMachine으로 이동

- Money 클래스의 0원 검증 및 로또 금액 배수 검증 로직을 LottoMachine으로 이동
- Money는 금액을 단순히 표현하는 역할에 집중하도록 개선 (음수 검증)
- 구매 검증 로직은 LottoMachine이 관리하여 도메인 간 책임을 명확하게 분리
- 로또 규칙 클래스의 로또 금액을 로또 머신으로 이동 (로또는 자신의 금액을 알 수 없기에 로또가 아닌 로또 머신으로 결정)

m-a-king committed:
test(Lotto): 로또 규칙 변경에 유연한 테스트로 수정

- 고정된 로또 규칙을 직접 사용하지 않고, LottoRule 값을 기반으로 동적으로 테스트
- 로또 번호 개수 및 범위가 변경되더라도 테스트가 자동으로 반영될 수 있도록 개선

m-a-king committed:
refactor: 불필요한 static 키워드 삭제, 코드 형식 정리

summerlunaa reviewed (approved):
강산 안녕하세요~
리뷰 반영 꼼꼼하게 잘 해주셨습니다! 리뷰마다 남겨주신 의견도 잘 읽었습니다.
이번 로또 미션은 여기서 마무리할게요. 첫 미션 진행하느라 너무 고생하셨습니다~! 일주일 동안 나눴던 대화들이 앞으로의 미션들에서도 도움이 되길 바랍니다 🚀 💯
